### PR TITLE
Update: add support for ecmaVersion 20xx (fixes #6750)

### LINKED ISSUE #6750: ES2015 aliases for ES6?
This is similar to #5169, but in a backwards-compatible way.

Basically, an ability to specify `ecmaVersion: 2015` that would behave just like `ecmaVersion: 6` without deprecating the value `6`.
And an `es2015` environment that would just be an alias for the `es6` one.

I understand if you think the maintenance overhead from this isn't worth it, but I wanted to check first :)

### PR DESCRIPTION
**Pull request by kaicataldo**
Thoughts on this approach? The way `verify()` works doesn't seem to make it easy to test this particular change. Is there something else I could spy on?

### CODE REVIEW

File: lib/eslint.js [hunk 1]
```diff
@@ -407,6 +407,30 @@ function isDisabledByReportingConfig(reportingConfig, ruleId, location) {
...
+ * Normalize ECMAScript version from the initial config
+ * @param  {number} ecmaVersion ECMAScript version from the initial config
+ * @param  {boolean} isModule Whether the source type is module or not
+ * @returns {number} normalized ECMAScript version
+ */
+function normalizeEcmaVersion(ecmaVersion, isModule) {
+
+    // need at least ES6 for modules
+    if (isModule && (!ecmaVersion || ecmaVersion < 6)) {
+        ecmaVersion = 6;
```
Comment (ilyavolodin) [inline lib/eslint.js]: This would be a breaking change. Right now, if you don't set es6 and turn on modules, nothing happens

Comment (kaicataldo) [inline lib/eslint.js]: Are you sure? I thought I was just moving the logic [from here](https://github.com/eslint/eslint/pull/6907/files#diff-b6559e6b095a292d244b6b7bae986291L466) into the normalize function

Comment (ilyavolodin) [inline lib/eslint.js]: Hmm.. Interesting. I though we didn't support that. But I might be wrong.

Comment (kaicataldo) [inline lib/eslint.js]: Seems like it! Thoughts on this approach if that's the case? Wasn't entirely sure about the tests in particular.

**Comment by ilyavolodin** [comment]
LGTM, but waiting another day for others to look

**Comment by alberto** [comment]
LGTM

File: lib/eslint.js [hunk 2]
```diff
@@ -407,6 +407,30 @@ function isDisabledByReportingConfig(reportingConfig, ruleId, location) {
...
+    }
+
+    if (ecmaVersion === 2016) {
```
Comment (nzakas) [inline lib/eslint.js]: I'd just say `if (ecmaVersion >= 2015) ecmaVersion -= 2009`. That way we don't have to keep updating this section.

File: lib/eslint.js (hunk 1 above)
Comment (nzakas) [inline lib/eslint.js]: This is correct. Setting to module implies ES6 when no ecmaVersion is present.

File: lib/eslint.js (hunk 2 above)
Comment (kaicataldo) [inline lib/eslint.js]: I like that - will update

Comment (platinumazure) [inline lib/eslint.js]: 2009 is a bit of a magic number here-- please consider including a comment with that change. Thanks!

Comment (kaicataldo) [inline lib/eslint.js]: Added comment as well - let me know if you think it can be improved

**Comment by kaicataldo** [comment]
Updated

File: tests/lib/eslint.js [hunk 3]
```diff
@@ -3116,15 +3112,50 @@ describe("eslint", function() {
...
+                it("the ECMASCript version number is 2016", function() {
+                    const messages = eslint.verify("x ** y;", {
+                        parserOptions: {
+                            ecmaVersion: 2016
+                        }
+                    });
+
+                    assert.equal(messages.length, 0);
+                });
+            });
```
Comment (btmills) [inline tests/lib/eslint.js]: Would it be worth it to add a test ensuring that it fails to parse the exponentiation operator when the version is 2015? If not, LGTM!

Comment (kaicataldo) [inline tests/lib/eslint.js]: I had considered that, and can definitely add that!

File: lib/eslint.js [hunk 4]
```diff
@@ -407,6 +407,27 @@ function isDisabledByReportingConfig(reportingConfig, ruleId, location) {
...
+ * @returns {number} normalized ECMAScript version
+ */
+function normalizeEcmaVersion(ecmaVersion, isModule) {
+
+    // need at least ES6 for modules
+    if (isModule && (!ecmaVersion || ecmaVersion < 6)) {
+        ecmaVersion = 6;
+    }
+
+    // calculate ECMAScript edition number from official year version
```
Comment (platinumazure) [inline lib/eslint.js]: I would love a comment explaining why 2009 specifically. Even just something like "(ES6 = ES2015, so subtract 2009)" would help a lot, in my view. But I don't feel that strongly about it.

Comment (kaicataldo) [inline lib/eslint.js]: Sure - updated. How does this look?

**Comment by kaicataldo** [comment]
Updated with suggested changes

**Comment by btmills** [comment]
LGTM, nice work @kaicataldo! Just waiting for tests to pass.

**Comment by platinumazure** [comment]
LGTM, thanks for considering my suggestion.

**Pull request merged**